python/src/scipp/_variable.py

```diff
@@ -50,24 +51,33 @@ def scalar(value: _Any,
       in which case type is inferred from value input.
       Cannot be specified for value types of
       str, Dataset or DataArray.
+    :param with_variance: Optional, boolean flag, if True includes variances
```

Comment:
`with_variance` and `variance` are not mutually exclusive. If `with_variance = False`, the variance is not stored even when passed in.

`sc.zero` would not cover the case of having a value but not a variance (yet). You could argue that one should never make an only partially initialised object and remove `with_variance`. The only real reason for having this argument is to preallocate memory and then fill it. But that does not matter for `scalar`.
But would you apply your comment to `array`, too?

Reply:
> with_variance and variance are not mutually exclusive. If with_variance = False, the variance is not stored even when passed in.

That sounds dangerous, do we really want to support that?

> sc.zero would not cover the case of having a value but not a variance (yet). You could argue that one should never make an only partially initialised object and remove with_variance. The only real reason for having this argument is to preallocate memory and then fill it. But that does not matter for scalar.

The only way to create something uninitialized should be `sc.empty` (or `sc.Variable`, which we want to discourage for direct use, eventually).

`sc.array` should *not* support `with_variance` imho, since it does not appear to have a reasonable meaning (unless you want to default or zero init).

Reply:
> That sounds dangerous, do we really want to support that?

Not sure. The alternative would be throwing.

Before this PR,  `sc.array` supports defaulted variances through the double meaning of the `variances` argument. Currently that is not possible as there is no `with_variance` argument.

The variances are never uninitialised, they are either the numbers you pass in or 0 (value initialised according to dtype).

Reply:
Yes, should throw if conflicting args are given. Also ensure that the creation functions like `sc.array` do not support calls that lead to conflicting args.

`sc.array` should *not* support defaulted variances (and should not support `with_variances`).